lib/index.js

```diff
@@ -508,124 +627,105 @@ Socket.prototype.send = function(msg, flags, cb) {
 
   if (Array.isArray(msg)) {
     for (var i = 0, len = msg.length; i < len; i++) {
-      var part = msg[i];
       var isLast = i === len - 1;
+      var msgFlags = isLast ? flags : flags | zmq.ZMQ_SNDMORE;
+      var callback = isLast ? cb : undefined;
 
-      if (!Buffer.isBuffer(part)) {
-        part = new Buffer(String(part), 'utf8');
-      }
-
-      this._outgoing.push([part, isLast ? flags : zmq.ZMQ_SNDMORE, isLast ? cb : undefined]);
+      this._outgoing.append(msg[i], msgFlags, callback);
     }
   } else {
-    if (!Buffer.isBuffer(msg)) {
-      msg = new Buffer(String(msg), 'utf8');
-    }
-
-    this._outgoing.push([msg, flags, cb]);
+    this._outgoing.append(msg, flags, cb);
   }
 
-  if (!(flags & zmq.ZMQ_SNDMORE)) {
-    try {
-      this._flush();
-    } catch (err) {
-      this.emit('error', err);
-    } finally {
-      this._zmq.pending = this._outgoing.length;
-    }
+  if (this._outgoing.canSend()) {
+    this._zmq.pending = true;
+    this._flushWrites();
+  } else {
+    this._zmq.pending = false;
   }
 
   return this;
 };
 
 
-// The workhorse that does actual send and receive operations.
-// This helper is called from `send` above, and in response to
-// the watcher noticing the signaller fd is readable.
-Socket.prototype._flush = function() {
-  var flags, args, emitArgs, errorSent;
-
-  // Don't allow recursive flush invocation as it can lead to stack
-  // exhaustion and write starvation
-  if (this._flushing) return;
-
-  this._flushing = true;
-  while (true) {
-    // If an Async thread that may call a zmq function is in flight
-    // do not call any zmq functions.
-    if (this._zmq.state !== zmq.STATE_READY) {
-      this._flushing = false;
-      return;
-    }
+Socket.prototype._flushRead = function () {
+  var message = this._zmq.readv(); // can throw
+  if (!message) {
+    return false;
+  }
 
-    flags = this._zmq.getsockopt(zmq.ZMQ_EVENTS);
+  // Handle received message immediately to prevent memory leak in driver
+  if (message.length === 1) {
+    // hot path
+    this.emit('message', message[0]);
+  } else {
+    this.emit.apply(this, ['message'].concat(message));
+  }
+  return true;
+};
 
-    if (!this._outgoing.length) flags &= ~zmq.ZMQ_POLLOUT;
-    if (!flags) {
-      this._flushing = false;
-      return;
-    };
+Socket.prototype._flushWrite = function () {
+  var batch = this._outgoing.fetch();
+  if (!batch) {
+    this._zmq.pending = false;
+    return false;
+  }
 
-    var emitArgs = this.read();
-    if (emitArgs) {
-      emitArgs.unshift('message');
-      // Handle received message immediately to prevent memory leak in driver
-      this.emit.apply(this, emitArgs);
+  try {
```

Comment:
Not having a try-catch block would lead to better performance. Would it be possible to redesign so it was not necessary?